firebase-firestore/src/main/java/com/google/firebase/firestore/local/QueryEngine.java

```diff
@@ -87,7 +102,48 @@ public ImmutableSortedMap<DocumentKey, Document> getDocumentsMatchingQuery(
       return result;
     }
 
-    return executeFullCollectionScan(query);
+    QueryContext context = new QueryContext();
+    result = executeFullCollectionScan(query, context);
+    if (result != null && automaticIndexingEnabled) {
+      createCacheIndexes(query, context, result.size());
+    }
+    return result;
+  }
+
+  /**
+   * Decides whether SDK should create a full matched field index for this query based on query
+   * context and query result size.
+   */
+  // TODO(csi): Auto experiment data.
+  private void createCacheIndexes(Query query, QueryContext context, int resultSize) {
+    if (context.getDocumentReadCount() < minCollectionSizeToAutoCreateIndex) {
+      Logger.debug(
+          LOG_TAG,
+          "SDK will only creates cache indexes for collection contains more than or equal to "
+              + "%s documents.",
+          minCollectionSizeToAutoCreateIndex);
+      return;
+    }
+
+    String decisionStr = "";
+    // If evaluation is updated, please update tests in SQLiteLocalStoreTest.java
+    if (context.getDocumentReadCount() > 2 * resultSize) {
+      indexManager.createTargetIndices(query.toTarget());
+    } else {
+      decisionStr = " not";
+    }
+
+    if (Logger.isDebugEnabled()) {
+      Logger.debug(
+          LOG_TAG,
+          "Query ran locally using a full collection scan, walking through %s documents in total "
```

Comment:
I think there is benefit to logging something when SDK decides not creating any indexes. Since developer may wonder why auto creation has no effect after they turn it on.